### PR TITLE
Feature/etcd3 support with tests

### PR DESCRIPTION
**Pull request by fbdlampayan**
etcd3-support feature tests. 
fully utilizes the ones available in the server_test.go for etcdv2.
Also included are bug fixes for the ones caught by the testcases.

### CODE REVIEW

File: backends/etcd3/etcd3.go [hunk 1]
```diff
@@ -16,6 +16,7 @@ import (
 	"fmt"
 	"github.com/coreos/etcd/mvcc/mvccpb"
 	"encoding/json"
+	//"github.com/skynetservices/skydns/backends/etcd"
```
Comment (miekg) [inline backends/etcd3/etcd3.go]: ?

File: backends/etcd3/etcd3.go [hunk 2]
```diff
@@ -41,11 +42,15 @@ func NewBackendv3 (client etcdv3.Client, ctx context.Context, config *Config) *B
 }
 
 func (g *Backendv3) Records(name string, exact bool) ([]msg.Service, error) {
+
```
Comment (miekg) [inline backends/etcd3/etcd3.go]: hmmm, whitespace is not but clutters this PR a bit.

File: backends/etcd3/etcd3.go [hunk 3]
```diff
@@ -74,25 +79,36 @@ func (g *Backendv3) ReverseRecord(name string) (*msg.Service, error) {
...
 			}
+
 			return r, e
 		}
 	})
 
 	if err != nil {
 		return nil, err
 	}
+	if resp == nil {
```
Comment (miekg) [inline backends/etcd3/etcd3.go]: why this? Isn't err != nil enough to bail out. Why not?

File: backends/etcd3/etcd3.go [hunk 4]
```diff
@@ -104,13 +120,38 @@ type bareService struct {
...
+ * etcdv3 doesn't treat its keys as directories anymore
+ * however skydns does, so this is needed to make the results
+ * from etcd3.get to respect the rules of skydns.
+ */
+func isItemDirTreeNode(nameParts []string, keyParts []string) bool {
+	for i, w := range nameParts {
+		if w == "*" || w == "any"{
+			return true;
+		}
+		if keyParts[i] != w && (!(w == "*" || w == "any")) {
```
Comment (miekg) [inline backends/etcd3/etcd3.go]: I find this hard to parse, can we just add another if statement?

File: backends/etcd3/etcd3.go [hunk 5]
```diff
@@ -104,13 +120,38 @@ type bareService struct {
...
+
+
 func (g *Backendv3) loopNodes(kv []*mvccpb.KeyValue, nameParts []string, star bool, bx map[bareService]bool) (sx []msg.Service, err error) {
 	if bx == nil {
 		bx = make(map[bareService]bool)
 	}
 Nodes:
 	for _, item := range kv {
 
+		s := string(item.Key[:])
```
Comment (miekg) [inline backends/etcd3/etcd3.go]: Why the `[:]`

File: backends/etcd3/etcd3errors.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+	Code int
+	Message string
+}
+
+func (e Etcd3Error) Error() string {
+	return fmt.Sprintf("%v - %v", e.Code, e.Message);
+}
+
+const (
+	KEYNOTFOUND = 100
```
Comment (miekg) [inline backends/etcd3/etcd3errors.go]: Can you add a comment where this value comes from?
Also, go nit, use CamelCase. And does it need to be exported?

File: server/server.go [hunk 7]
```diff
@@ -20,6 +20,7 @@ import (
 	etcd "github.com/coreos/etcd/client"
 	"github.com/coreos/go-systemd/activation"
 	"github.com/miekg/dns"
+	etcd3 "github.com/skynetservices/skydns/backends/etcd3"
```
Comment (miekg) [inline server/server.go]: please sort this with the other skydns imports.

File: server/server.go [hunk 8]
```diff
@@ -877,14 +886,32 @@ func isTCP(w dns.ResponseWriter) bool {
 	return ok
 }
 
-// etcNameError return a NameError to the client if the error
-// returned from etcd has ErrorCode == 100.
-func isEtcdNameError(err error, s *server) bool {
+func verdictEtcdV2(err error, s *server) bool {
```
Comment (miekg) [inline server/server.go]: IsEtcd2NameError?

File: server/server.go [hunk 9]
```diff
@@ -877,14 +886,32 @@ func isTCP(w dns.ResponseWriter) bool {
...
 		return true
 	}
 	if err != nil {
-		logf("error from backend: %s", err)
+		logf("error from backend v2: %s", err)
 	}
 	return false
 }
+
+func verdictEtcdV3(err error, s *server) bool {
```
Comment (miekg) [inline server/server.go]: isEtcd3NameError?

File: backends/etcd3/etcd3errors.go (hunk 6 above)
Comment (fbdlampayan) [inline backends/etcd3/etcd3errors.go]: The value is arbitrary, so to respect on what skydns has been using with etcdv2 I used the value of 100.

File: backends/etcd3/etcd3.go (hunk 2 above)
Comment (fbdlampayan) [inline backends/etcd3/etcd3.go]: okay will remove the whitespaces a bit

File: backends/etcd3/etcd3.go (hunk 1 above)
Comment (fbdlampayan) [inline backends/etcd3/etcd3.go]: arrgh, yea my bad, missed deleting that one.

File: server/server.go (hunk 7 above)
Comment (fbdlampayan) [inline server/server.go]: Ahhh noticed just now, it is sorted alphabetically. Okay will do

**Comment by fbdlampayan** [comment]
updated with cleanup based on comments.
